File: src/rac/services/parallel_build.py
"""Parallel cold build of the derived read-model (ADR-107).

The cold build's cost is dominated by parsing — a fresh whole-corpus walk spends
roughly three-quarters of its wall time in :func:`rac.core.markdown.parse_file`
(the markdown tokenise + frontmatter parse), the rest in deriving the index /
relationship / token structures and serialising the segment store. Parsing is
embarrassingly parallel and pure per file, so this module fans it out across
processes while leaving the derive and serialise phases exactly as the serial
path runs them.

**Determinism is the contract, not a nicety.** Workers each parse a *contiguous*
range of the sorted ``find_markdown_files`` list, and the parent concatenates the
ranges back in list order, so the parsed snapshot is byte-for-byte the same
sequence :func:`rac.core.corpus.walk_corpus` yields — regardless of how many
workers ran. The whole derive/serialise pipeline downstream is a pure function of
that ordered snapshot, so the store bytes and every served response are identical
across worker counts (the ADR-107 worker-invariance rule). A test hashes the
segment files of a ``workers=1`` and a ``workers=4`` build and asserts equality.

**Workers call the one true parse path.** ``_worker_parse`` imports and calls the
same :func:`parse_file` / :func:`classify` the serial walk does — never a
reimplementation — so every pinned parse behaviour (``errors="replace"`` lossy
decode, the BOM-defeats-frontmatter rule, the byte-cap oversize issue) is
reproduced exactly. The byte cap is read from ``RAC_MAX_FILE_BYTES`` on every
call, and spawned workers inherit the environment, so a lowered cap applies in
workers and serial alike.

**Correctness never depends on the parallel rung.** Below a measured file-count
threshold, or on a box with ``cpu_count() <= 2``, the build stays single-process
(forking costs more than it saves at small N). Any worker fault — an exception,
a crashed child, a pickling failure — is caught and the build falls back to the
serial :func:`walk_corpus`, which can never produce a partial or corrupt
snapshot. The parallel path is a latency lever, never a correctness dependency
(the same posture ADR-080 takes for the cache).

Stdlib only (``multiprocessing`` with the spawn context — no forked-state or
lambda closures cross the boundary; the worker is a module-level function).
"""

from __future__ import annotations

import multiprocessing as mp
import os
import sys
import time
from dataclasses import dataclass
from pathlib import Path

from rac.core.classification import classify
from rac.core.corpus import CorpusEntry, walk_corpus
from rac.core.fs import find_markdown_files
from rac.core.markdown import parse_file
from rac.services.derived_cache import DerivedIndex, build_derived_index_from_entries

_TIMING_ENV = "RAC_TIMING"

# Spawn-safe fault-injection hook. When this env var is set, every worker raises
# on entry — exercising the worker-crash -> serial-fallback path in a *real*
# subprocess (a parent-side monkeypatch cannot reach a spawned child, which
# re-imports this module fresh). The env is inherited across spawn, so setting it
# in a test makes the children fault. Never set in production.
_FAULT_ENV = "RAC_PARALLEL_BUILD_FAULT"

# Below this file count the spawn + IPC overhead outweighs the parse win, so the
# cold build stays single-process. Measured crossover on the 4-core reference
# node: a serial parse of ~5k small artifacts is a couple of seconds, and the
# spawn of a worker set plus the round-trip of the parsed snapshot eats most of
# the theoretical speedup below that. Above it the parse win dominates. Tunable
# via RAC_PARALLEL_BUILD_MIN_FILES for measurement; never lowers correctness.
DEFAULT_MIN_PARALLEL_FILES = 5_000
_MIN_FILES_ENV = "RAC_PARALLEL_BUILD_MIN_FILES"


@dataclass
class BuildStats:
    """Per-phase cold-build timings for the ``RAC_TIMING`` scorecard line.

    ``workers`` is the number of worker processes the parse actually used — ``1``
    means the single-process path ran (small corpus, ``cpu_count() <= 2``, or a
    worker fault fell back). ``write_ms`` is filled in by the caller that owns the
    store write (the cache or the compaction path), which is where serialisation
    happens; the build itself leaves it ``0``.
    """

    files: int
    workers: int
    parse_ms: float = 0.0
    derive_ms: float = 0.0
    write_ms: float = 0.0


def _min_parallel_files() -> int:
    raw = os.environ.get(_MIN_FILES_ENV)
    if raw is None:
        return DEFAULT_MIN_PARALLEL_FILES
    try:
        value = int(raw)
    except ValueError:
        return DEFAULT_MIN_PARALLEL_FILES
    return value if value >= 0 else DEFAULT_MIN_PARALLEL_FILES


def _resolve_workers(workers: int | None, n_files: int) -> int:
    """How many worker processes to use — 1 means run single-process.

    An explicit ``workers`` (the tests' worker-count-invariance lever) is honoured
    up to the file count, so a small corpus can still be built with 4 workers to
    prove determinism. With ``workers=None`` the default policy applies: stay
    single-process on a 1-2 core box or below the file-count threshold, otherwise
    use every core.
    """
    if n_files <= 1:
        return 1
    if workers is not None:
        return max(1, min(workers, n_files))
    cpu = os.cpu_count() or 1
    if cpu <= 2 or n_files < _min_parallel_files():
        return 1
    return min(cpu, n_files)


def _contiguous_chunks(items: list[Path], n: int) -> list[list[Path]]:
    """Split ``items`` into at most ``n`` contiguous, order-preserving ranges.

    Contiguity is the determinism guarantee: concatenating the ranges in list
    order reproduces the original sorted sequence exactly, so the merged snapshot
    is worker-count-invariant.
    """
    if n <= 1:
        return [items]
    size = (len(items) + n - 1) // n
    return [items[i : i + size] for i in range(0, len(items), size)]


def _worker_parse(paths: list[Path]) -> list[CorpusEntry]:
    """Parse a contiguous range of paths through the one true core parse path.

    Identical per-file work to :func:`rac.core.corpus.walk_corpus`:
    :func:`parse_file` then :func:`classify`, wrapped in a :class:`CorpusEntry`
    carrying the original :class:`~pathlib.Path`. No reimplementation — the pinned
    decode / BOM / oversize semantics come from ``parse_file`` verbatim.
    """
    if os.environ.get(_FAULT_ENV):
        raise RuntimeError("parallel-build worker fault (injected)")
    entries: list[CorpusEntry] = []
    for path in paths:
        product = parse_file(str(path))
        entries.append(CorpusEntry(path=path, product=product, classification=classify(product)))
    return entries


def _parse_serial(paths: list[Path]) -> list[CorpusEntry]:
    # The serial fallback parses directly (not via the fault-gated ``_worker_parse``)
    # so it succeeds even when the fault env var is set: the crash-resilience test
    # sets it to make the *workers* fault, then expects this single-process path to
    # complete. Same per-file work as ``walk_corpus`` / ``_worker_parse``.
    entries: list[CorpusEntry] = []
    for path in paths:
        product = parse_file(str(path))
        entries.append(CorpusEntry(path=path, product=product, classification=classify(product)))
    return entries


def _parse_parallel(paths: list[Path], n_workers: int) -> list[CorpusEntry] | None:
    """Fan the parse out across ``n_workers`` spawned processes, or ``None`` on fault.

    Returns the parsed entries in sorted-path order, or ``None`` if any worker
    fault occurred — the caller then falls back to the serial path. Uses the spawn
    context so no forked interpreter state and no closure crosses the boundary.
    """
    chunks = _contiguous_chunks(paths, n_workers)
    try:
        ctx = mp.get_context("spawn")
        with ctx.Pool(len(chunks)) as pool:
            results = pool.map(_worker_parse, chunks)
    except BaseException:
        # Any child fault (exception, crash, pickling failure) collapses to the
        # serial path. A partially-mapped result is discarded whole — the store is
        # never written from a truncated snapshot.
        return None
    return [entry for chunk in results for entry in chunk]


def parallel_parse_paths(
    paths: list[Path], *, workers: int | None = None
) -> tuple[list[CorpusEntry], int]:
    """Parse an explicit list of paths, parallel when it pays; entries in list order.

    Returns ``(entries, workers_used)``. ``workers_used == 1`` signals the
    single-process path ran (threshold, core count, or a fault fallback). The
    entry order matches ``paths`` exactly, so a caller passing sorted paths gets a
    sorted snapshot regardless of worker count.
    """
    n_workers = _resolve_workers(workers, len(paths))
    if n_workers <= 1:
        return _parse_serial(paths), 1
    entries = _parse_parallel(paths, n_workers)
    if entries is None:
        return _parse_serial(paths), 1
    return entries, n_workers


def parallel_walk(
    directory: str, *, recursive: bool = True, workers: int | None = None
) -> tuple[list[CorpusEntry], int]:
    """Parse a whole corpus, parallel when it pays; entries in sorted-path order.

    The parallel analogue of ``list(walk_corpus(directory))``: byte-identical
    output (same entries, same sorted order), only fanned across processes. When
    the single-process path runs it delegates to :func:`walk_corpus` so the two
    code paths share one definition of the serial walk.
    """
    paths = find_markdown_files(directory, recursive=recursive)
    n_workers = _resolve_workers(workers, len(paths))
    if n_workers <= 1:
        return list(walk_corpus(directory, recursive=recursive)), 1
    entries = _parse_parallel(paths, n_workers)
    if entries is None:
        return list(walk_corpus(directory, recursive=recursive)), 1
    return entries, n_workers


def build_derived_index_parallel(
    directory: str, *, recursive: bool = True, workers: int | None = None
) -> tuple[DerivedIndex, BuildStats]:
    """Build the derived read-model with a parallel parse, then the serial derive.

    Byte-identical to :func:`rac.services.derived_cache.build_derived_index`: the
    parse is fanned out but produces the same ordered snapshot, and the derive
    phase (:func:`build_derived_index_from_entries`) is the same pure function of
    that snapshot. Returns the derived index plus the per-phase timings the
    ``RAC_TIMING`` scorecard reports; ``write_ms`` is filled by the caller that
    serialises the store.
    """
    t0 = time.perf_counter()
    entries, used = parallel_walk(directory, recursive=recursive, workers=workers)
    t1 = time.perf_counter()
    derived = build_derived_index_from_entries(directory, entries, recursive=recursive)
    t2 = time.perf_counter()
    return derived, BuildStats(
        files=len(entries),
        workers=used,
        parse_ms=(t1 - t0) * 1000.0,
        derive_ms=(t2 - t1) * 1000.0,
    )


def emit_build_timing(stats: BuildStats) -> None:
    """Write the cold-build scorecard line to stderr when ``RAC_TIMING`` is set.

    Env-gated and stderr-only (stdout is a frozen contract); absent by default.
    Mirrors the incremental-validate ``rac-timing:`` line shape (ADR-106).
    """
    if _TIMING_ENV not in os.environ:
        return
    sys.stderr.write(
        f"rac-timing: build_parse_ms={stats.parse_ms:.3f} "
        f"build_derive_ms={stats.derive_ms:.3f} build_write_ms={stats.write_ms:.3f} "
        f"workers={stats.workers} files={stats.files}\n"
    )

File: rust/rac-engine/src/parallel_build.rs
//! Parallel cold build of the derived read-model (ADR-107/ADR-108).
//!
//! Native shape of `services/parallel_build.py` + `parallel_merge.py`: the
//! per-document work (parse, classify, index row, token vectors, live/scope
//! projection) fans out across rayon workers as compact *fragments*; the
//! parent merge runs only the cross-document steps (graph resolution,
//! inbound fill, portfolio) — in sorted-path order, so the store bytes are
//! worker-count-invariant by construction (no pickling boundary exists
//! in-process; ADR-114).
//!
//! Correctness never depends on the parallel rung: below the file-count
//! threshold, on a 1–2 core box, or on ANY worker fault (a panic in a
//! fragment task — exercised by `RAC_PARALLEL_BUILD_FAULT`), the build
//! falls back to the authoritative serial floor, whose partial results are
//! never written.

use std::path::PathBuf;

use crate::derived::{build_derived_index_from_items, DerivedIndex, DECISION_TYPE};
use crate::relationships::{relationships_from_corpus, CorpusItem};
use crate::resolve::{entry_from_item, field_tokens_of, is_live_decision, IndexEntry};
use crate::retrieve::{scope_rows_from_items, ScopeRow};

const TIMING_ENV: &str = "RAC_TIMING";
/// Fault-injection hook: when set, every fragment task panics — exercising
/// the fault → serial-floor degrade in a real parallel run. Never set in
/// production.
const FAULT_ENV: &str = "RAC_PARALLEL_BUILD_FAULT";
/// Below this file count the fan-out's coordination overhead outweighs the
/// win, so the cold build stays on the serial floor (the oracle's measured
/// crossover, kept for contract fidelity).
pub const DEFAULT_MIN_PARALLEL_FILES: usize = 5_000;
const MIN_FILES_ENV: &str = "RAC_PARALLEL_BUILD_MIN_FILES";

/// Per-phase cold-build timings for the `RAC_TIMING` scorecard line.
#[derive(Default)]
pub struct BuildStats {
    pub files: usize,
    pub workers: usize,
    pub parse_ms: f64,
    pub derive_ms: f64,
    pub write_ms: f64,
}

fn min_parallel_files() -> usize {
    let Some(raw) = std::env::var_os(MIN_FILES_ENV) else {
        return DEFAULT_MIN_PARALLEL_FILES;
    };
    match raw.to_string_lossy().trim().parse::<i64>() {
        Ok(value) if value >= 0 => value as usize,
        _ => DEFAULT_MIN_PARALLEL_FILES,
    }
}

/// How many workers to use — 1 means the serial floor. An explicit count
/// (the worker-invariance lever) is honoured up to the file count; the
/// default policy stays serial on a small box or below the threshold.
fn resolve_workers(workers: Option<usize>, n_files: usize) -> usize {
    if n_files <= 1 {
        return 1;
    }
    if let Some(w) = workers {
        return w.max(1).min(n_files);
    }
    let cpu = std::thread::available_parallelism()
        .map(std::num::NonZeroUsize::get)
        .unwrap_or(1);
    if cpu <= 2 || n_files < min_parallel_files() {
        return 1;
    }
    cpu.min(n_files)
}

/// One document's compact derived projection — the unit workers emit.
struct DocFragment {
    item: CorpusItem,
    index_entry: IndexEntry, // inbound stays 0; the merge fills it
    field_tokens: crate::resolve::FieldTokens,
    is_live_decision: bool,
    scope_row: Option<ScopeRow>,
}

fn fragment_for(path_display: &str) -> DocFragment {
    if std::env::var_os(FAULT_ENV).is_some() {
        panic!("parallel-build worker fault (injected)");
    }
    let artifact = crate::parse::parse_file(path_display);
    let spec = crate::spec::spec_for(&crate::classify::classify(&artifact).artifact_type);
    let item = CorpusItem {
        path: path_display.to_string(),
        artifact,
        spec,
    };
    let index_entry = entry_from_item(&item, 0);
    let field_tokens = field_tokens_of(&index_entry);
    let live = item.spec.map(|s| s.name == DECISION_TYPE).unwrap_or(false)
        && is_live_decision(&item.artifact);
    let scope_row = scope_rows_from_items(std::slice::from_ref(&item)).into_iter().next();
    DocFragment {
        item,
        index_entry,
        field_tokens,
        is_live_decision: live,
        scope_row,
    }
}

/// Fan the parse + per-doc derive across `n_workers`, or None on any fault.
fn fragments_parallel(paths: &[String], n_workers: usize) -> Option<Vec<DocFragment>> {
    let pool = rayon::ThreadPoolBuilder::new()
        .num_threads(n_workers)
        .build()
        .ok()?;
    let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
        pool.install(|| {
            use rayon::prelude::*;
            paths
                .par_iter()
                .map(|path| fragment_for(path))
                .collect::<Vec<DocFragment>>()
        })
    }));
    // A partially-derived result is discarded whole — the store is never
    // written from a faulted fan-out.
    result.ok()
}

/// Reproduce the derived read-model from per-document fragments: only the
/// cross-document steps run here, in sorted-path order.
fn reproduce(fragments: Vec<DocFragment>, directory: &str, recursive: bool) -> DerivedIndex {
    let mut items = Vec::with_capacity(fragments.len());
    let mut index_entries = Vec::with_capacity(fragments.len());
    let mut field_tokens = Vec::with_capacity(fragments.len());
    let mut live_decision_paths = Vec::new();
    let mut scope_rows = Vec::new();
    for fragment in fragments {
        if fragment.is_live_decision {
            live_decision_paths.push(fragment.index_entry.path.clone());
        }
        if let Some(row) = fragment.scope_row {
            scope_rows.push(row);
        }
        items.push(fragment.item);
        index_entries.push(fragment.index_entry);
        field_tokens.push(fragment.field_tokens);
    }
    // The cross-document steps: graph resolution, inbound fill, portfolio.
    let relationships = relationships_from_corpus(&items);
    let mut inbound: std::collections::HashMap<&str, i64> = std::collections::HashMap::new();
    for rel in &relationships {
        if let Some(resolved) = &rel.resolved_path {
            *inbound.entry(resolved.as_str()).or_insert(0) += 1;
        }
    }
    for entry in &mut index_entries {
        entry.inbound_count = inbound.get(entry.path.as_str()).copied().unwrap_or(0);
    }
    let summary = crate::portfolio::portfolio_from_corpus(directory, &items, recursive);
    DerivedIndex {
        index_entries,
        field_tokens,
        relationships,
        live_decision_paths,
        portfolio_summary: crate::output::portfolio_summary_value(&summary),
        scope_rows,
    }
}

/// Build the derived read-model with a parallel parse AND per-doc derive —
/// byte-identical to `derived::build_derived_index` for any worker count.
pub fn build_derived_index_parallel(
    directory: &str,
    recursive: bool,
    workers: Option<usize>,
) -> (DerivedIndex, BuildStats) {
    let t0 = std::time::Instant::now();
    let paths: Vec<String> = crate::walk::find_markdown_files(directory, recursive)
        .into_iter()
        .map(|e| e.display)
        .collect();
    let n_workers = resolve_workers(workers, paths.len());
    let fragments = if n_workers > 1 {
        fragments_parallel(&paths, n_workers)
    } else {
        None
    };
    if let Some(fragments) = fragments {
        let used = n_workers;
        let t1 = std::time::Instant::now();
        let n_files = fragments.len();
        let derived = reproduce(fragments, directory, recursive);
        let t2 = std::time::Instant::now();
        return (
            derived,
            BuildStats {
                files: n_files,
                workers: used,
                parse_ms: (t1 - t0).as_secs_f64() * 1000.0,
                derive_ms: (t2 - t1).as_secs_f64() * 1000.0,
                write_ms: 0.0,
            },
        );
    }
    // Serial floor: the authoritative walk + derive; a fault above lands here.
    let items = crate::relationships::corpus_items(directory, recursive);
    let t1 = std::time::Instant::now();
    let derived = build_derived_index_from_items(directory, &items, recursive);
    let t2 = std::time::Instant::now();
    (
        derived,
        BuildStats {
            files: items.len(),
            workers: 1,
            parse_ms: (t1 - t0).as_secs_f64() * 1000.0,
            derive_ms: (t2 - t1).as_secs_f64() * 1000.0,
            write_ms: 0.0,
        },
    )
}

/// Write the cold-build scorecard line to stderr when `RAC_TIMING` is set —
/// env-gated, stderr-only, byte-shaped like the oracle's (ADR-107).
pub fn emit_build_timing(stats: &BuildStats) {
    if std::env::var_os(TIMING_ENV).is_none() {
        return;
    }
    eprintln!(
        "rac-timing: build_parse_ms={:.3} build_derive_ms={:.3} build_write_ms={:.3} workers={} files={}",
        stats.parse_ms, stats.derive_ms, stats.write_ms, stats.workers, stats.files
    );
}

/// The paths type alias for the freshness tracker's explicit-list parse
/// (INDEX-PLAN B6): parse a known path list through the one true per-file
/// path, parallel when it pays; entries in list order.
pub fn parallel_parse_paths(paths: &[PathBuf]) -> (Vec<CorpusItem>, usize) {
    let displays: Vec<String> = paths
        .iter()
        .map(|p| p.to_string_lossy().into_owned())
        .collect();
    use rayon::prelude::*;
    let items: Vec<CorpusItem> = displays
        .par_iter()
        .map(|path| {
            let artifact = crate::parse::parse_file(path);
            let spec =
                crate::spec::spec_for(&crate::classify::classify(&artifact).artifact_type);
            CorpusItem {
                path: path.clone(),
                artifact,
                spec,
            }
        })
        .collect();
    let workers = std::thread::available_parallelism()
        .map(std::num::NonZeroUsize::get)
        .unwrap_or(1);
    (items, workers)
}
